### ISSUE TITLE
Warning when using mimetype validator if no file is set with form input

### ISSUE THREAD

**Issue by M-pire** [davidyell/CakePHP3-Proffer#28]
Followed your project description for the plugin. I have the "mimetype" validator rule set for the UsersTable. When using a test form (type=file) with input (type=file) and fieldname "image", I get the following warning:

>Warning (2): finfo::file(): Empty filename or path [ROOT\vendor\davidyell\proffer\src\Model\Validation\ProfferRules.php, line 70]

which boils down to the following code:
> $type = $finfo->file($value['tmp_name'], FILEINFO_MIME_TYPE);

in the mimetype() method of the class.

I think there should be an isset() and empty()  just above it. If empty/null/not exists, then return false as well.

p.s. Thanks for the plugin! Great work

**Comment by davidyell** [comment]
Thanks, I'll get this fixed.

**Comment by davidyell** [comment]
Hey @M-pire did you also set `->allowEmpty()` in your validation?

**Comment by M-pire** [comment]
@davidyell Sorry for late reaction; at that time I think I didn't have the validator allowEmpty added for the field. At the moment I do have it set and that seems to work. Thanks!

I'm struggling a little to get the filename/dir saved to the database ( it saves "Array" into "image" database field currently ). I'll get it right someday hehe.

**Comment by davidyell** [comment]
Hey @M-pire, that might be because the plugin isn't loaded, or that the behaviour isn't attached the the table class.

https://github.com/davidyell/CakePHP3-Proffer#configuration

**Comment by M-pire** [comment]
@davidyell : I have added the plugin in bootstrap.php, like this:
```php
Plugin::load('Proffer');
```

then added/customized the behavior for UsersTable:
```php
   $this->addBehavior('Proffer.Proffer', [
        		'image' => [   
	        		'root' => WWW_ROOT . 'files', 
	        		'dir' => 'image_dir',  
	        		'thumbnailSizes' => [
	        			'thumb' => ['w' => 200, 'h' => 200]
	        		],
	        		'thumbnailMethod' => 'imagick'
        		]
        ]);
```

In my database, there is a "users" table with the fields mentioned in the behavior: "image" and "image_dir". As said before, at the moment the table field "image" is getting "Array" as content after save. 

My form looks like this:
```php
echo $this->Form->create($user, ["novalidate", "type" => "file"]);
// couple other fields
echo $this->Form->input('image', ["label" => "Avatar", 'type' => 'file']);
// other fields and submit
```

I have tried to test if it might be "imagick" failing, but I don't see any errors in the logs and tested Imagick on other projects without problems ( on the same server )

In UsersTable the validator includes the Proffer validation, but no rules at the moment for testing:

```php
    	$validator->provider('proffer', 'Proffer\Model\Validation\ProfferRules');
        $validator->allowEmpty('image', 'update')
// and some other default validator rules, but no specific Proffer rules yet
```

I'll try to find the problem, but thought you might see an obvious one here :)
Thanks in advance.

**Comment by M-pire** [comment]
I have done some more testing. 
1. Removed the plugin load in bootstrap.php, it shows me "Proffer plugin not found", so I assume it knows/loads it
2. Added a "test()" method to your ProfferBehavior class and tested that "$this->Users->test()" works. It did, so I assume it correctly loads the plugin
3. I recently updated CakePHP3 to RC2 in my app, so figured maybe the update might have changed something? I see they done some work on events, but are only deprecated and not removed yet.

You any ideas? I'll test some more, to see if the callbacks (afterX) are called.

**Comment by davidyell** [comment]
4) The behaviours `beforeSave` method not being called might be because your Table has a `beforeSave()` method which isn't calling it's parent or returning true.

If your beforeSave's don't call the parent or don't return true it can stop the events chain from firing which would then prevent the behaviour's method from being run.

**Comment by M-pire** [comment]
I didn't think about that, so nice catch :) I'm afraid my UsersTable has no beforeFilter . I do have a User entity (which only contains _accessible => "*" and _setPassword() for hashing).

I'll try the plugin in a clean project and see if i can get it working there. 

Thanks for your help so far!

**Comment by davidyell** [comment]
Not `beforeFilter()` but `beforeSave()`

**Comment by M-pire** [comment]
Oops, i meant beforeSave() :) The UsersTable at this moment only contains these methods: initialize, validationDefault($validator) and buildRules($rules).

**Comment by davidyell** [comment]
I assume that you've checked that the name of the field in your database, such as `photo` or `image` matches the field used in your Form?

**Comment by M-pire** [comment]
yes, my eyes hurt checking it again and again. I thought I might have developed tunnel vision, so I looked again today. I'm still sure it's something stupid somewhere. I'll find it, someday :)

**Comment by M-pire** [comment]
I've tried a fresh install of only CakePHP3-RC2 and your plugin (both installed via composer). I've baked a Users controller/model/view and added form type=>file and the "image"-field (type=file). In the database for "users" I have both "image" (varchar 200, not null) and "image_dir" (varchar 200, not null). I have added no other code, besides loading the plugin in bootstrap.php . 

This still gives me "Array" in the database after selecting an image and submitting the form. When i don't add the fields to the database, the files do get uploaded correctly. I have no idea why it's not working.
